Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1.  Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ * 2.  Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "SwitchTrackMac.h"
+
+#if PLATFORM(MAC)
+
+#import "ControlFactoryMac.h"
+#import "FloatRoundedRect.h"
+#import "GraphicsContext.h"
+#import "LocalCurrentGraphicsContext.h"
+#import "LocalDefaultSystemAppearance.h"
+#import "SwitchTrackPart.h"
+#import <pal/spi/mac/CoreUISPI.h>
+#import <pal/spi/mac/NSAppearanceSPI.h>
+
+namespace WebCore {
+
+SwitchTrackMac::SwitchTrackMac(SwitchTrackPart& part, ControlFactoryMac& controlFactory, WebSwitchView *switchPtr)
+    : ControlMac(part, controlFactory)
+    , m_switchPtr(switchPtr)
+{
+}
+
+FloatRect SwitchTrackMac::rectForBounds(const FloatRect& bounds, const ControlStyle& style) const
+{
+    ControlStyle cs(style);
+    auto rect = bounds;
+    rect.inflateY(-1/ 2);
+    return rect;
+}
+
+void SwitchTrackMac::draw(GraphicsContext& context, const FloatRoundedRect& borderRect, float deviceScaleFactor, const ControlStyle& style)
+{
+    LocalDefaultSystemAppearance localAppearance(style.states.contains(ControlStyle::State::DarkAppearance), style.accentColor);
+
+    bool isOn = style.states.contains(ControlStyle::State::Checked);
+    bool isEnabled = style.states.contains(ControlStyle::State::Enabled);
+    bool isPressed = style.states.contains(ControlStyle::State::Pressed);
+    bool isInActiveWindow = style.states.contains(ControlStyle::State::WindowActive);
+
+    auto inflatedRect = rectForBounds(borderRect.rect(), style);
+
+    // FIXME: _trackFrameInRect relies on interface definition in SwitchThumbMac
+    NSRect trackRect = [[m_switchPtr nativeSwitch] _trackFrameInRect:inflatedRect];
```

Comment:
I don't know how to improve this.

Reply:
The shared header I mentioned above is the correct thing to do here.